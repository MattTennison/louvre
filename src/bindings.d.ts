export {}

declare global {
  const PEXELS_API_KEY: string
  const PHOTOS: KVNamespace
}
