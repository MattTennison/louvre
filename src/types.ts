import { KVNamespace } from '@cloudflare/workers-types'

export type Env = {
  PHOTOS: KVNamespace
  PEXELS_API_KEY: string
}
