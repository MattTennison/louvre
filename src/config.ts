export default {
  pexels: {
    authentication: 'foo',
  },
}
