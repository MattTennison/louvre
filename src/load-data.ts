import { search } from './services/photo'

export async function handleScheduled() {
  const results = await search('minimalism')

  const individualPhotoItems = results.photos.map((photo) => ({
    key: `minimalism:photo:${photo.id}`,
    value: photo,
  }))

  const itemsToWrite = [
    ...individualPhotoItems,
    {
      key: 'minimalism:photo:list',
      value: {
        keys: results.photos.map((photo) => ({
          name: `minimalism:photo:${photo.id}`,
        })),
      },
    },
  ]

  await Promise.all([
    ...itemsToWrite.map(({ key, value }) =>
      PHOTOS.put(key, JSON.stringify(value)),
    ),
  ])
}
