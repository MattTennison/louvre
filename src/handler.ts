import { pickRandom } from './utils/pick-random'
import cronParser from 'cron-parser'
import config from './config'
import { PexelsPhotoPayload } from './services/photo'

const createResponse = (body: Object, init?: ResponseInit) => {
  const response = new Response(JSON.stringify(body), init)
  response.headers.set('Access-Control-Allow-Origin', '*')
  response.headers.set('Content-Type', 'application/json')
  return response
}

export async function handleRequest(request: Request): Promise<Response> {
  try {
    const { keys } = await PHOTOS.list({ prefix: 'minimalism:photo' })
    if (keys.length === 0) {
      const response = createResponse(
        { error: 'no.photos.in.cache' },
        {
          status: 503,
        },
      )

      response.headers.set(
        'Retry-After',
        cronParser
          .parseExpression(config.schedule.cron, { tz: 'Europe/London' })
          .next()
          .toISOString(),
      )

      return response
    }

    const { name } = pickRandom(keys)
    const result = await PHOTOS.get(name)

    if (result === null) {
      return createResponse(
        { error: 'missing.photo.in.cache' },
        {
          status: 500,
        },
      )
    }

    const photo = PexelsPhotoPayload.parse(JSON.parse(result))
    const {
      src: { original },
      avg_color,
      photographer,
      photographer_url,
      url,
    } = photo

    return createResponse({
      data: {
        photo: {
          src: original,
          avg_color,
          attribution: {
            photographer: { name: photographer, url: photographer_url },
            link: { source: 'Pexels', url },
          },
        },
      },
    })
  } catch (e) {
    return createResponse(
      { error: e.message },
      {
        status: 500,
      },
    )
  }
}
