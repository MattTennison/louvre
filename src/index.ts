import { handleRequest } from './handler'
import { handleScheduled } from './load-data'

export default {
  fetch: () => handleRequest(),
  scheduled: () => handleScheduled(),
}
