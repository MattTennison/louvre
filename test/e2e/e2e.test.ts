import pexelsSearchPhotosResponse from '../fixtures/pexels/search-photos.json'
import { KeyValueStore, stubKeyValueStore } from '../stub-key-value-store'
import { handleScheduled } from '../../src/load-data'
import { handleRequest } from '../../src/handler'

declare var global: any
let photosStore: KeyValueStore

beforeEach(() => {
  photosStore = stubKeyValueStore()
  jest.spyOn(Math, 'random').mockReturnValue(0.64)

  const today = '2021-08-22T09:14:47.409Z'
  jest.spyOn(Date, 'now').mockReturnValue(new Date(today).getTime())

  Object.assign(global, {
    PHOTOS: photosStore,
    fetch: () =>
      Promise.resolve({
        json: () => Promise.resolve(pexelsSearchPhotosResponse),
      }),
  })
  jest.resetModules()
})

test('with initalised store', async () => {
  await handleScheduled()

  const response = await handleRequest(new Request('/', { method: 'GET' }))
  const body = await response.json()

  expect(response.status).toBe(200)
  expect(response.headers.get('Content-Type')).toEqual('application/json')
  expect(response.headers.get('Access-Control-Allow-Origin')).toEqual('*')
  expect(body).toMatchInlineSnapshot(`
Object {
  "avg_color": "#BFC0C5",
  "url": "https://images.pexels.com/photos/1699655/pexels-photo-1699655.jpeg",
}
`)
})

test('when KeyValue is throwing errors', async () => {
  await handleScheduled()
  jest
    .spyOn(photosStore, 'get')
    .mockRejectedValue(new Error('something.not.right'))

  const response = await handleRequest(new Request('/', { method: 'GET' }))
  const body = await response.json()

  expect(response.status).toEqual(500)
  expect(body).toMatchInlineSnapshot(`
Object {
  "error": "something.not.right",
}
`)
})

test('when no entries are in the KeyValue store', async () => {
  const response = await handleRequest(new Request('/', { method: 'GET' }))
  const body = await response.json()

  expect(response.status).toEqual(503)
  expect(response.headers.get('Retry-After')).toEqual(
    '2021-08-22T23:30:00.000Z',
  )
  expect(body).toMatchInlineSnapshot(`
Object {
  "error": "no.photos.in.cache",
}
`)
})
