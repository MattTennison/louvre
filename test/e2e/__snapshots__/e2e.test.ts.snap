// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when KeyValue is throwing errors: 500 Error Response Body 1`] = `
Object {
  "error": "something.not.right",
}
`;

exports[`when no entries are in the KeyValue store: 503 Error Response Body 1`] = `
Object {
  "error": "no.photos.in.cache",
}
`;

exports[`with initalised store: 200 Success Response Body 1`] = `
Object {
  "data": Object {
    "photo": Object {
      "attribution": Object {
        "link": Object {
          "source": "Pexels",
          "url": "https://www.pexels.com/photo/white-wooden-cabinet-1699655/",
        },
        "photographer": Object {
          "name": "Henry & Co.",
          "url": "https://www.pexels.com/@hngstrm",
        },
      },
      "avg_color": "#BFC0C5",
      "src": "https://images.pexels.com/photos/1699655/pexels-photo-1699655.jpeg",
    },
  },
}
`;
